oxenss/storage/database.cpp

```diff
@@ -817,19 +819,22 @@ std::pair<std::vector<message>, bool> Database::retrieve(
     }
 
     auto st = impl->prepared_st(
-            last_id ? "SELECT hash, namespace, timestamp, expiry, data FROM messages "
-                      "WHERE owner = ? AND namespace = ? AND id > ? ORDER BY id LIMIT ?"
-                    : "SELECT hash, namespace, timestamp, expiry, data FROM messages "
-                      "WHERE owner = ? AND namespace = ? ORDER BY id LIMIT ?");
+            "SELECT hash, namespace, timestamp, expiry, data FROM messages "
+            "WHERE owner = ? AND namespace = ? {} ORDER BY id {} LIMIT ?"_format(
+                    last_id ? "AND id > ?" : "", reverse_direction ? "DESC" : "ASC"));
```

Comment:
I'm not 100% sure of this, but is this working properly if both `last_id` and `reverse_direction` are true?  I *think* when both are specified the `AND id > ?` should get flipped to `AND id < ?`.